### PR TITLE
adds nixosGenerate function

### PR DESCRIPTION
**Pull request by mayl**
Feeling a little out of my depth, but an attempt to address #94, which is similar to some use cases I have.  An example of how I'd like to use this [here](https://github.com/mayl/generatorFlakes). 

Several things about this implementation seem really hacky, particularly using a `callPackage` call purely to extract the `formatAttr` value out of a format file.  Review, comments and suggestions of other approaches are much appreciated!

### CODE REVIEW

**Comment by mayl** [comment]
@Lassulus @mkg20001 can either of you take a look at this and comment on if this looks like it's on the right track?

**Comment by Lassulus** [comment]
well I'm just getting started with flakes. But once I've gotten a bit more knowledgeable I can take a look. maybe @Mic92 wants to take a look :D

File: flake.nix [hunk 1]
```diff
@@ -10,13 +10,27 @@
...
 
   # Library modules (depend on nixlib)
-  {
+  rec {
     # export all generator formats in ./formats
     nixosModules = nixlib.lib.mapAttrs' (file: _: {
       name = nixlib.lib.removeSuffix ".nix" file;
       # The exported module should include the internal format* options
       value.imports = [ (./formats + "/${file}") ./format-module.nix ];
+      # !!! TODO: is there a better way to extract formatAttr for the format?
```
Comment (mkg20001) [inline flake.nix:19]: yes.

check ./format-module.nix, it makes "formatAttr" an option

that means you can use config.formatAttr instead

```
let
  system = nixpkgs.lib.nixosSystem { ... };
in
  system.config.system.build.${system.config.formatAttr}
```

Comment (mayl) [reply flake.nix]: Thanks for pointing out the options interface to `nixos-generators` I had glossed over that.  Take a look at my latest commit.

**Comment by mayl** [comment]
> well I'm just getting started with flakes. But once I've gotten a bit more knowledgeable I can take a look. maybe @Mic92 wants to take a look :D

Well, even so you almost certainly already know more than I do. I think in this case it's really as simple as trying to write a `nix` function which evaluates to a `nixos-generate` image, and then assigning that function to an attribute under `outputs` for users to consume.

**Comment by mayl** [comment]
I haven't heard any input that this is heading in the wrong direction, and after implementing @mkg20001's suggested change I don't have any other implementation feedback so taking this out of draft.  

@Lassulus @Mic92 are there more changes required or does this look ready to merge to you?

File: flake.nix [hunk 2]
```diff
@@ -10,13 +10,26 @@
...
-  {
+  rec {
     # export all generator formats in ./formats
     nixosModules = nixlib.lib.mapAttrs' (file: _: {
       name = nixlib.lib.removeSuffix ".nix" file;
       # The exported module should include the internal format* options
       value.imports = [ (./formats + "/${file}") ./format-module.nix ];
     }) (builtins.readDir ./formats);
+
+    nixosGenerate = { system, pkgs, modules, format }:
```
Comment (Mic92) [inline flake.nix:21]:
```suggestion
    # Example usage in flakes:
    # outputs = { self, nixpkgs, nixos-generators, ...}:  {
    #   vmware = nixos-generators.nixosGenerate { 
    #     pkgs = nixpkgs.legacyPackages.x86_64-linux;
    #     modules = [./configuration.nix];
    #     format = "vmware";
    #   };
    # }
    nixosGenerate = { pkgs, modules, format }:
```

File: flake.nix [hunk 3]
```diff
@@ -10,13 +10,26 @@
...
+    let 
+      formatModule = builtins.getAttr format nixosModules;
+      image = nixpkgs.lib.nixosSystem {
+        inherit system pkgs;
```
Comment (Mic92) [inline flake.nix:25]:
```suggestion
        inherit pkgs;
        system = pkgs.system;
```

File: flake.nix (hunk 2 above)
Comment (Mic92) [reply flake.nix:21]: We don't need `system` as it is already implied by `pkgs`.

**Comment by Mic92** [comment]
An example in the README would be nice.

File: flake.nix (hunk 2 above)
Comment (mayl) [reply flake.nix:21]: Thanks looks good

**Comment by mayl** [comment]
Thanks for the input @Mic92 - very helpful.  Either GH doesn't allow it because I don't have write access to this repo, or I'm being really dumb but I couldn't find the "accept proposed changes" button on any of your changes so I re-entered them in my branch to eliminate the `system` parameter and comment the function like you did.  I also added an example of using this new feature to the `readme`. Please feel free to take a look and provide any additional feedback.

File: flake.nix [hunk 4]
```diff
@@ -10,13 +10,35 @@
...
-  {
+  rec {
     # export all generator formats in ./formats
     nixosModules = nixlib.lib.mapAttrs' (file: _: {
       name = nixlib.lib.removeSuffix ".nix" file;
       # The exported module should include the internal format* options
       value.imports = [ (./formats + "/${file}") ./format-module.nix ];
     }) (builtins.readDir ./formats);
+
+    # example useage in flakes:
```
Comment (Mic92) [inline flake.nix:21]:
```suggestion
    # example usage in flakes:
```

File: README.md [hunk 5]
```diff
@@ -90,6 +90,39 @@ For more details about this have a look at :
...
+    pkgs = nixpkgs.legacyPackages.x86_64-linux;
+    modules = [./configuration.nix];
+    format = "vmware";
+  };
+  vbox = nixos-generators.nixosGenerate {
+    pkgs = nixpkgs.legacyPackages.x86_64-linux;
+    modules = [./configuration.nix];
+    format = "virtualbox";
+  };
+}
```
Comment (Mic92) [inline README.md:123]: The flake syntax is not quite correct here:

```suggestion
{
  inputs = {
    nixpkgs.url = "nixpkgs/nixos-unstable";
    nixos-generators = {
      url = "github:nix-community/nixos-generators";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };
  outputs = { self, nixpkgs, nixos-generators, ... }: {
    packages.x86_64-linux = {
      vmware = nixos-generators.nixosGenerate {
        pkgs = nixpkgs.legacyPackages.x86_64-linux;
        modules = [
          # you can include your own nixos configuration here, i.e.
          # ./configuration.nix
        ];
        format = "vmware";
      };
      vbox = nixos-generators.nixosGenerate {
        pkgs = nixpkgs.legacyPackages.x86_64-linux;
        modules = [];
        format = "virtualbox";
      };
    };
  };
}
```

**Comment by mayl** [comment]
I added your suggested fix.  For what it's worth, in the [example](https://github.com/mayl/generatorFlakes/blob/master/flake.nix) I've been testing with, I don't have the `packages.x86_64-linux` super attribute and it seems to just pretty much work fine.  When is that attribute necessary?

**Comment by Mic92** [comment]
> I added your suggested fix. For what it's worth, in the [example](https://github.com/mayl/generatorFlakes/blob/master/flake.nix) I've been testing with, I don't have the `packages.x86_64-linux` super attribute and it seems to just pretty much work fine. When is that attribute necessary?

`nix build` tries to prefix `packages.<arch>.<packagename>` before `packages.<packagename>`. If you leave it out you will get issues once you try to support different architectures/operating systems. Also some tools like `nix flake check` will complain about invalid schemas.

**Comment by mayl** [comment]
> > I added your suggested fix. For what it's worth, in the [example](https://github.com/mayl/generatorFlakes/blob/master/flake.nix) I've been testing with, I don't have the `packages.x86_64-linux` super attribute and it seems to just pretty much work fine. When is that attribute necessary?
> 
> `nix build` tries to prefix `packages.<arch>.<packagename>` before `packages.<packagename>`. If you leave it out you will get issues once you try to support different architectures/operating systems. Also some tools like `nix flake check` will complain about invalid schemas.

Got it, thanks for that clarification.

I've been thinking of two improvements.  First would be to go through and implement all the command line script's arguments in this function (it's currently missing stuff like `format_dir` etc.).  Once that's done, we could then have the script build through this function (obviously there would be a little refactoring here so you wouldn't have to use a `flakes` enabled `nix build`).  Once that was done, there would be one common path which builds the images regardless of which way you use `nixos-generators`.  Any thoughts on these ideas?

However, if you think it's ok I'm inclined to try and merge this minimal feature set before all that, and tackle that in a separate PR.  But I'm not trying to rush it in - just trying to keep making progress.  What do you think, is this getting close to being merge worthy?  What are the next steps?

File: flake.nix [hunk 6]
```diff
@@ -10,13 +10,35 @@
...
+
+    # example usage in flakes:
+    #   outputs = { self, nixpkgs, nixos-generators, ...}: {
+    #     vmware = nixos-generators.nixosGenerate {
+    #       pkgs = nixpkgs.legacyPackages.x86_64-linux;
+    #       modules = [./configuration.nix];
+    #       format = "vmware";
+    #   };
+    # }
+    nixosGenerate = { pkgs, modules, format }:
```
Comment (Mic92) [inline flake.nix:29]: I think we can merge this after this nitpick:

```suggestion
    nixosGenerate = { pkgs, format,  modules ? [ ]}:
```

File: README.md [hunk 7]
```diff
@@ -90,6 +90,46 @@ For more details about this have a look at :
...
+        pkgs = nixpkgs.legacyPackages.x86_64-linux;
+        modules = [
+          # you can include your own nixos configuration here, i.e.
+          # ./configuration.nix
+        ];
+        format = "vmware";
+      };
+      vbox = nixos-generators.nixosGenerate {
+        pkgs = nixpkgs.legacyPackages.x86_64-linux;
+        modules = [];
```
Comment (Mic92) [inline README.md:125]:
```suggestion
```

**Comment by mayl** [comment]
Nitpicks welcome, suggestions applied.

**Pull request merged**